perceval/components/experiment.py

```diff
@@ -0,0 +1,767 @@
+            # Remove output ports used to connect the new experiment
+            for i in mode_mapping:
+                port = self.get_output_port(i)
+                if port is not None:
+                    del self._out_ports[port]
+
+        perm_modes, perm_component = ModeConnector.generate_permutation(mode_mapping)
+        if perm_component is not None:
+            self._components.append((perm_modes, perm_component))
+
+        sorted_modes = tuple(range(min(mode_mapping), min(mode_mapping) + component.m))
+        self._components.append((sorted_modes, component))
+        self._is_unitary = self._is_unitary and isinstance(component, ACircuit)
+        self._has_td = self._has_td or isinstance(component, TD)
+
+    def _add_herald(self, mode, expected, name=None):
+        """
+        This internal implementation neither increases the herald count nor decreases the mode of interest count
+        """
+        if not self.are_modes_free([mode], PortLocation.IN_OUT):
+            raise UnavailableModeException(mode, "Another port overlaps")
+        if name is None:
+            name = self._anon_herald_num
+            self._anon_herald_num += 1
+        self._in_ports[Herald(expected, name)] = [mode]
+        self._out_ports[Herald(expected, name)] = [mode]
+        self._mode_type[mode] = ModeType.HERALD
+        self._circuit_changed()
+
+    def add_herald(self, mode: int, expected: int, name: str = None) -> Self:
+        r"""
+        Add a heralded mode
+
+        :param mode: Mode index of the herald
+        :param expected: number of expected photon as input AND output on the given mode (must be 0 or 1)
+        :param name: Herald port name. If none is passed, the name is auto-generated
+        """
+        assert expected == 0 or expected == 1, "expected must be 0 or 1"
+        self._add_herald(mode, expected, name)
+        self._n_moi -= 1
+        self._n_heralds += 1
+        return self
+
+    @property
+    def components(self):
+        return self._components
+
+    @property
+    def m(self) -> int:
+        """
+        :return: Number of modes of interest (MOI) defined in the experiment
+        """
+        return self._n_moi
+
+    @m.setter
+    def m(self, value: int):
+        if self._n_moi != 0:
+            raise RuntimeError(f"The number of modes of this experiment was already set (to {self._n_moi})")
+        if not isinstance(value, int) or value < 1:
+            raise ValueError(f"The number of modes should be a strictly positive integer (got {value})")
+        self._n_moi = value
+        self._detectors = [None] * value
+        self._mode_type = [ModeType.PHOTONIC] * value
+
+    @property
+    def circuit_size(self) -> int:
+        r"""
+        :return: Total size of the enclosed circuit (i.e. self.m + heralded mode count)
+        """
+        return self._n_moi + self._n_heralds
+
+    def linear_circuit(self, flatten: bool = False, use_phase_noise=False) -> Circuit:
```

Comment:
`linear` is not the correct term here, it was kept for compatibility.
Since `Experiment` is a new class, we should rename it to `unitary_circuit` (and adapt the docstring accordingly)